pkg/controller/volume/expand/expand_controller.go

```diff
@@ -123,107 +108,151 @@ func NewExpandController(
-			return
-		}
-		pvc, ok = tombstone.Obj.(*v1.PersistentVolumeClaim)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("tombstone contained object that is not a pvc %#v", obj))
-			return
-		}
+func (expc *expandController) syncHandler(key string) error {
+	namespace, name, err := kcache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	pvc, err := expc.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		klog.V(5).Infof("Error getting PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+
+	pv, err := getPersistentVolume(pvc, expc.pvLister)
+	if err != nil {
+		klog.V(5).Infof("Error getting Persistent Volume for PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
 	}
 
-	expc.resizeMap.DeletePVC(pvc)
+	volumeSpec := volume.NewSpecFromPersistentVolume(pv, false)
+	volumePlugin, err := expc.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
+	if err != nil || volumePlugin == nil {
+		msg := fmt.Errorf("didn't find a plugin capable of expanding the volume; " +
+			"waiting for an external controller to process this PVC")
+		expc.recorder.Event(pvc, v1.EventTypeNormal, events.ExternalExpanding,
+			fmt.Sprintf("Ignoring the PVC: %v.", msg))
+		klog.V(3).Infof("Ignoring the PVC %q (uid: %q) : %v.",
+			util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, msg)
+		return err
+	}
+
+	return expc.expand(pvc, pv)
 }
 
-func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
-	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+func (expc *expandController) expand(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) error {
+	pvc, err := util.MarkResizeInProgress(pvc, expc.kubeClient)
+	if err != nil {
+		klog.V(5).Infof("Error setting PVC %s in progress with error : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), err)
+		return err
+	}
 
-	if oldPVC == nil || !ok {
-		return
+	generatedOperations, err := expc.operationGenerator.GenerateExpandVolumeFunc(pvc, pv)
+	if err != nil {
+		klog.Errorf("Error starting ExpandVolume for pvc %s with %v", util.GetPersistentVolumeClaimQualifiedName(pvc), err)
+		return err
 	}
+	klog.V(5).Infof("Starting ExpandVolume for volume %s", util.GetPersistentVolumeClaimQualifiedName(pvc))
+	expand := func() (eventErr, detailedErr error) {
+		// Handle unhandled panics (very unlikely)
+		defer runtime.HandleCrash()
+		if generatedOperations.CompleteFunc != nil {
+			defer generatedOperations.CompleteFunc(&detailedErr)
+		}
+		if generatedOperations.EventRecorderFunc != nil {
+			defer generatedOperations.EventRecorderFunc(&eventErr)
+		}
+		// Handle panic, if any, from operationFunc()
+		defer runtime.RecoverFromPanic(&detailedErr)
+		return generatedOperations.OperationFunc()
+	}
+	_, detailedErr := expand()
```

Comment:
It sounds like `GeneratedOperations` needs Run function which can be used to execute enqueued operation and gather metrics, handle crash etc.

Reply:
added